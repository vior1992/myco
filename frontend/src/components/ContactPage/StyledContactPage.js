import styled from 'styled-components';

export const StyledContactPageWrapper = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  margin: 50px 10px 0 10px;
  text-align: center;

  h1 {
    font-family: 'Architects Daughter', cursive;
  }

  @media (min-width: 420px) {
    h1 {
      font-size: ${(props) => props.theme.sizes['xx-huge']};
    }
  }
`;

export const StyledForm = styled.form`
  font-family: 'Play', sans-serif;
  background-color: #0b1325;
  justify-content: space-around;
  border-radius: ${(props) => props.theme.sizes['x-small']};
  box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2);
  transition: 0.3s;
  padding: ${(props) => props.theme.sizes.small};
  margin: ${(props) => props.theme.sizes['x-small']};

  :hover {
    box-shadow: 0 8px 16px 0 rgba(0, 0, 0, 0.2);
  }

  button {
    background-color: transparent;
    border: 0;
    padding: 0;
  }

  i {
    font-size: ${(props) => props.theme.sizes['x-large']};
    color: white;
    margin: ${(props) => props.theme.sizes.medium};
    padding: ${(props) => props.theme.sizes.medium};
  }

  @media (min-width: 420px) {
    width: 500px;
    align-self: center;
    min-height: 500px;
    font-size: ${(props) => props.theme.sizes['x-large']};
    padding: ${(props) => props.theme.sizes['x-large']};

    i {
      font-size: ${(props) => props.theme.sizes['xx-huge']};
    }
  }
`;

export const StyledInputWrapper = styled.div`
  display: flex;
  flex-direction: column;
  margin: ${(props) => props.theme.sizes['x-small']}; 0;

  label {
    color: white;
    text-align: start;
    margin: ${(props) => props.theme.sizes['x-small']} 0;
  }
  input {
    background-color: transparent;
    border: 0;
    border-bottom: 1px solid white;
    color: white;
  }

  p {
    color: #de0e0e;
    font-size: ${(props) => props.theme.sizes.small};
  }

  @media (min-width: 420px) {
    input {
      font-size: ${(props) => props.theme.sizes.large};
    }

    p {
      font-size: ${(props) => props.theme.sizes.medium};
    }
  }
`;
