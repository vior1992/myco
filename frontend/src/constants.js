const PROPS_TO_DISPLAY = [
  { key: 'first_name', title: 'First name' },
  { key: 'last_name', title: 'Last name' },
  { key: 'email', title: 'Email' },
  { key: 'phone_number', title: 'Phone number' },
];

const MOCKED_CONTACTS = [
  {
    id: 1,
    first_name: 'Ragnar',
    last_name: 'Lodbrok',
    email: 'vikings@mocked.com',
    phone_number: '+ 00 00 01',
  },
  {
    id: 2,
    first_name: 'Bj√∂rn',
    last_name: 'Ironside',
    email: 'vikings@mocked.com',
    phone_number: '+ 00 00 02',
  },
];

export { PROPS_TO_DISPLAY, MOCKED_CONTACTS };
