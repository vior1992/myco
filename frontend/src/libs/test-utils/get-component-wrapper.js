import React from 'react';
import { ToastProvider } from 'react-toast-notifications';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import { GlobalStyles, Theme } from '../../styles';

const getComponentWrapper = ({
  Component,
  mockedProps = {},
  initialRoute = '/',
}) => {
  const history = createMemoryHistory({ initialEntries: [initialRoute] });

  return (
    <Theme>
      <GlobalStyles />
      <ToastProvider>
        <Router history={history}>
          <Component {...mockedProps} />
        </Router>
      </ToastProvider>
    </Theme>
  );
};

export default getComponentWrapper;
