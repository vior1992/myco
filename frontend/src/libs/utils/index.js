import buildRequestOptions from './build-request-options';

export { buildRequestOptions };
