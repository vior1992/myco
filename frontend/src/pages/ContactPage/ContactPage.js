import React, { useEffect, useState } from 'react';
import { useHistory, Link } from 'react-router-dom';
import { useToasts } from 'react-toast-notifications';
import {
  StyledContactPageWrapper,
  StyledForm,
  StyledInputWrapper,
} from './StyledContactPage';
import { contactsRepository } from '../../repositories';
import { buildRequestOptions } from '../../libs/utils';
import {
  CONTACTS_PATH,
  CONTACT_PAGE_ICONS,
  CONTACT_PAGE_TITLES,
  INITIAL_CONTACT_DATA,
  PROPS_TO_DISPLAY,
} from '../../constants';

const { ADD_ICON, EDIT_ICON } = CONTACT_PAGE_ICONS;
const { ADD_TITLE, EDIT_TITLE } = CONTACT_PAGE_TITLES;

const ContactPage = () => {
  const [contactData, setContactData] = useState(INITIAL_CONTACT_DATA);
  const [errors, setErrors] = useState({});
  const history = useHistory();
  const { addToast } = useToasts();

  const [, id = ''] = history.location.pathname.split(CONTACTS_PATH);
  const submitIconClass = `fas fa-${id ? EDIT_ICON : ADD_ICON}`;
  const pageTitle = id ? EDIT_TITLE : ADD_TITLE;

  useEffect(() => {
    if (id) {
      const fetchData = async () => {
        const data = await contactsRepository.getContact({
          id,
          addToast,
          history,
        });

        setContactData(data);
      };

      fetchData();
    }
  }, [id, addToast, history]);

  const handleOnChange = ({ target: { id, value } }) => {
    setContactData({ ...contactData, [id]: value });
  };

  const handleOnSubmit = async (e) => {
    e.preventDefault();

    const options = buildRequestOptions(id, contactData);

    const isSuccessfulRequest = await contactsRepository.updateContact({
      id,
      options,
      addToast,
      setErrors,
    });

    isSuccessfulRequest && history.push('/');
  };

  return (
    <StyledContactPageWrapper>
      <h1>{pageTitle}</h1>
      <StyledForm onSubmit={handleOnSubmit}>
        {PROPS_TO_DISPLAY.map(({ key, title }) => (
          <StyledInputWrapper key={key}>
            <label htmlFor={key}>{title}</label>
            <input
              type="text"
              id={key}
              value={contactData[key]}
              onChange={handleOnChange}
            />
            {errors[key] && <p>{errors[key][0]}</p>}
          </StyledInputWrapper>
        ))}
        <div>
          <Link to="/">
            <i className="fas fa-arrow-left" />
          </Link>
          <button type="submit">
            <i className={submitIconClass} />
          </button>
        </div>
      </StyledForm>
    </StyledContactPageWrapper>
  );
};

export default ContactPage;
