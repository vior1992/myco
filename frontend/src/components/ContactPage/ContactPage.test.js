import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import ContactPage from './ContactPage';
import { MOCKED_CONTACTS } from '../../constants';
import { getComponentWrapper } from '../../libs/test-utils';

const setup = (initialRoute) => {
  const wrappedComponent = getComponentWrapper({
    Component: ContactPage,
    initialRoute,
  });
  const _render = render(wrappedComponent);
  return { ..._render };
};

describe('ContactPage component', () => {
  beforeAll(() => jest.spyOn(window, 'fetch'));

  const onSubmit = jest.fn((e) => e.preventDefault);

  it('should render without throwing an error & match snapshot', () => {
    const { asFragment } = setup('/contacts');
    expect(asFragment()).toMatchSnapshot();
  });

  it('should display error message if on submit request params not fil validations', async () => {
    const expectedValue = 'mocked test error';
    window.fetch.mockResolvedValueOnce({
      status: 422,
      json: async () => ({
        first_name: [expectedValue],
      }),
    });

    setup('/contacts');

    await waitFor(() => {
      const submit = screen.getByRole('button');
      fireEvent.click(submit);
    });

    const error = screen.getByText(expectedValue);
    expect(error).toBeInTheDocument();
  });

  it('should trigger handleOnChange function when input type', () => {
    const expectedValue = 'component test';

    setup('/contacts');

    const input = screen.getByRole('textbox', { name: 'First name' });
    expect(input.value).toEqual('');

    fireEvent.change(input, {
      target: {
        value: expectedValue,
      },
    });

    expect(input.value).toEqual(expectedValue);
  });

  it('should trigger handleOnSubmit function when submit', async () => {
    window.fetch.mockResolvedValueOnce({
      json: async () => MOCKED_CONTACTS[0],
    });
    window.fetch.mockResolvedValueOnce({
      json: async () => true,
    });

    setup('/contacts/1');

    waitFor(() => {
      const submit = screen.getByRole('button');
      fireEvent.click(submit);

      expect(onSubmit).toHaveBeenCalled();
    });

    fetch.mockClear();
  });
});
