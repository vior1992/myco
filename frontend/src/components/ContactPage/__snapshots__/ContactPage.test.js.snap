// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactPage component should render without throwing an error & match snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-bdnylx iFJVhQ"
  >
    <h1>
      ADD CONTACT
    </h1>
    <form
      class="sc-gtssRu WGoOk"
    >
      <div
        class="sc-dlnjPT cCFnob"
      >
        <label
          for="first_name"
        >
          First name
        </label>
        <input
          id="first_name"
          type="text"
          value=""
        />
      </div>
      <div
        class="sc-dlnjPT cCFnob"
      >
        <label
          for="last_name"
        >
          Last name
        </label>
        <input
          id="last_name"
          type="text"
          value=""
        />
      </div>
      <div
        class="sc-dlnjPT cCFnob"
      >
        <label
          for="email"
        >
          Email
        </label>
        <input
          id="email"
          type="text"
          value=""
        />
      </div>
      <div
        class="sc-dlnjPT cCFnob"
      >
        <label
          for="phone_number"
        >
          Phone number
        </label>
        <input
          id="phone_number"
          type="text"
          value=""
        />
      </div>
      <div>
        <a
          href="/"
        >
          <i
            class="fas fa-arrow-left"
          />
        </a>
        <button
          type="submit"
        >
          <i
            class="fas fa-user-plus"
          />
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
